openvdb_ax/openvdb_ax/ax.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+#include "ax.h"
+#include "ast/AST.h"
+#include "compiler/Compiler.h"
+#include "compiler/PointExecutable.h"
+#include "compiler/VolumeExecutable.h"
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace ax {
+
+/// @note Implementation for initialize, isInitialized and unitialized
+///       reamins in compiler/Compiler.cc
+
+void run(const char* ax, openvdb::GridBase& grid)
+{
+    // Construct a generic compiler
+    openvdb::ax::Compiler compiler;
+    // Parse the provided code and produce an abstract syntax tree
+    // @note  Throws with parser errors if invalid. Parsable code does not
+    //        necessarily equate to compilable code
+    const openvdb::ax::ast::Tree::ConstPtr ast = openvdb::ax::ast::parse(ax);
+
+    if (grid.isType<points::PointDataGrid>()) {
+        // Compile for Point support and produce an executable
+        // @note  Throws compiler errors on invalid code. On success, returns
+        //        the executable which can be used multiple times on any inputs
+        const openvdb::ax::PointExecutable::Ptr exe =
+            compiler.compile<openvdb::ax::PointExecutable>(*ast);
+        // Execute on the provided points
+        // @note  Throws on invalid point inputs such as mismatching types
+        exe->execute(static_cast<points::PointDataGrid&>(grid));
+    }
+    else {
+        // Compile for numerical grid support and produce an executable
+        // @note  Throws compiler errors on invalid code. On success, returns
+        //        the executable which can be used multiple times on any inputs
+        const openvdb::ax::VolumeExecutable::Ptr exe =
+            compiler.compile<openvdb::ax::VolumeExecutable>(*ast);
+        // Execute on the provided numerical grid
+        // @note  Throws on invalid grid inputs such as mismatching types
+        exe->execute(grid);
+    }
+}
+
+void run(const char* ax, openvdb::GridPtrVec& grids)
+{
+    if (grids.empty()) return;
+    // Check the type of all grids. If they are all points, run for point data.
```

Comment:
I know this is not an easy thing to solve, but this logic seems weird to me. You have an array of point grids and accidentally add a density grid on the end and it totally changes the way the other grids are processed. Also, what happens if you want to add a volumesample() operation? Wouldn't you have to adjust this logic?

Reply:
I've switched this to throw if the vector contains both points and volumes - as you say, when this functionality becomes available it'll change the result